File: ClientApp/src/app/shared/project.ts
import { Blueprint, SubComponent } from './Blueprint';
import { ITaggable } from './TagFilteredList';
import { v4 as uuidv4 } from 'uuid';


export class Project implements ITaggable {

  constructor(p?: Project, b?: Blueprint) {
    if (b !== undefined && b !== null) {
      this.Blueprint = b.Id;
      this.Components = b.Components.map((c: SubComponent) => new ProjectComponent(c));
    } else if (p !== undefined && p !== null) {
      this.Id = p.Id;
      this.Tags = p.Tags;
      this.Name = p.Name;
      this.Blueprint = p.Blueprint;
      this.Notes = p.Notes;
      this.Components = p.Components;
      this.Owner = p.Owner;
      this.Description = p.Description;
      this.DateCreated = p.DateCreated;
    }
    this.Display = true;
  }
  Id : string;
  Tags: string[];
  Display: boolean;
  Name: string;
  Blueprint: string;
  Notes: string;
  Components: ProjectComponent[];
  Owner: string;
  Description: string;
  DateCreated: string;
}


export class ProjectComponent {

  constructor(s?: SubComponent) {
    if (s !== undefined) {
      this.BaseComponent = s.Id;
      this.NeedsReplacing = false;
      this.Notes = '';
      this.Code = this.GenerateCode();
    }
  }
  ID: string;
  BaseComponent: string;
  NeedsReplacing: boolean;
  Notes: string;
  Code: string;


  private GenerateCode(): string {
    return uuidv4().substring(0, 8);
  }
}

File: Models/data/Project.cs
namespace Bluprintr.Models.Data
{
    using System;
    using System.Collections.Generic;
    using Bluprintr.Models.Data.Interfaces;
    using MongoDB.Bson;
    using MongoDB.Bson.Serialization.Attributes;

    /// <summary>A user-created project</summary>
    /// <seealso cref="Bluprintr.Models.Data.Interfaces.IMongoElement" />
    public class Project : IMongoElement
    {
        /// <summary>Gets or sets the identifier.</summary>
        /// <value>The ID of the project</value>
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string Id { get; set; }

        /// <summary>Gets or sets the Name.</summary>
        /// <value>The Name of the project</value>
        public string Name { get; set; }

        /// <summary>Gets or sets the notes.</summary>
        /// <value>The notes attached to the project</value>
        public string Notes { get; set; }

        /// <summary>Gets or sets the components.</summary>
        /// <value>The components that the project contains</value>
        public List<ProjectComponent> Components { get; set; }

        /// <summary>Gets or sets the owner.</summary>
        /// <value>The owner of the project</value>
        public string Owner { get; set; }

        /// <summary>Gets or sets whether the project publically accessible</summary>
        /// <value>Whether the project is publically accessible</value>
        public bool? IsPublic { get; set; }

        /// <summary>Gets or sets the Blueprint.</summary>
        /// <value>The Blueprint of the project</value>
        public string Blueprint { get; set; }

        /// <summary>Gets or sets the Description.</summary>
        /// <value>The Description of the project</value>
        public string Description { get; set; }

        /// <summary>Gets or sets the Created Date.</summary>
        /// <value>The date the project was created</value>
        public string DateCreated { get; set; }

        /// <summary>Gets or sets the tags</summary>
        /// <value>The tags attached to the project</value> 
        public string Tags { get; set; }
    }
}
